File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ancestor = -1;
    void diff(TreeNode* node,int minE,int maxE) {
        if(!node) return;
        
        minE = min(minE,node->val);
        maxE = max(maxE,node->val);
        
        ancestor = max(ancestor,abs(maxE-minE));
        
        diff(node->left,minE,maxE);
        diff(node->right,minE,maxE);
    }
    int maxAncestorDiff(TreeNode* root) {
        diff(root,root->val,root->val);
        return ancestor;
    }
};

File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ancestor = -1;
    public void diff(TreeNode node,int max, int min) {
        if(node == null) return;
        
        max = Math.max(max, node.val);
        min = Math.min(min,node.val);
        
        ancestor = Math.max(ancestor,Math.abs(max-min));
        diff(node.left,max,min);
        diff(node.right,max,min);
        
    }
    public int maxAncestorDiff(TreeNode root) {
        diff(root,root.val,root.val);
        return ancestor;
    }
}